### PR TITLE
Added Vector, Matrix and SparseMatrix classes to the math API

### PR DESCRIPTION
**Pull request by whaeck**

### CODE REVIEW

**Comment by whaeck** [comment]
This pull request requires the pull request on the Eigen library to be resolved. I am closing this one until then.

**Comment by whaeck** [comment]
Reopened the request now that the Eigen pull request has been merged.

No need to merge it yet, this is just to see if the test are run properly now.

**Comment by whaeck** [comment]
This can be merged now. I'll work on the operator*= issue with the SParseMatrix when I have time.

File: API/Array.hpp.in [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```
Comment (apmccartney) [inline API/Array.hpp.in]: generally speaking, classes defined by `typedef` or `using` statement can live in namespace header, but to err on the side of too fine a granularity is better than to err on the side of too coarse a granularity

**Comment by apmccartney** [comment]
If this feature is complete, you have my blessing to merge

File: API/Array.hpp.in (hunk 1 above)
Comment (whaeck) [inline API/Array.hpp.in]: True. Having these typedefs/usings set up in separate files will facilitate things when we want to replace them later on. Everything is set up as if they were concrete classes.
The unit library will be similar in setup but on a larger scale.

**Pull request merged**